server/src/main/java/io/littlehorse/common/model/getable/core/wfrun/ThreadRunModel.java

```diff
@@ -213,6 +215,75 @@ public NodeRunModel getCurrentNodeRun() {
         return getNodeRun(currentNodePosition);
     }
 
+    public void validateVariablesAndStart(Map<String, VariableValueModel> variables) {
+        if (currentNodePosition > 0) {
+            throw new IllegalStateException("Should only be called on creation");
+        }
+        currentNodePosition = 0;
+
+        Date now = new Date();
+        ThreadSpecModel threadSpec = getThreadSpecModel();
+        setStatus(LHStatus.RUNNING);
+        setStartTime(now);
+
+        NodeModel entrypointNode = threadSpec.getNodes().get(threadSpec.getEntrypointNodeName());
+
+        NodeRunModel entrypointRun = new NodeRunModel(processorContext);
+        entrypointRun.setThreadRun(this);
+        entrypointRun.setNodeName(entrypointNode.name);
+        entrypointRun.setStatus(LHStatus.STARTING);
+        entrypointRun.setId(new NodeRunIdModel(wfRun.getId(), this.number, 0));
+        entrypointRun.setWfSpecId(wfSpecId);
+        entrypointRun.setThreadSpecName(threadSpecName);
+        entrypointRun.setArrivalTime(now);
+        entrypointRun.setSubNodeRun(entrypointNode.getSubNode().createSubNodeRun(now));
+        putNodeRun(entrypointRun);
+
+        try {
+            threadSpec.validateStartVariables(variables);
+        } catch (LHValidationError exn) {
+            log.error("Invalid variables received", exn);
+            // TODO: determine how observability events should look like for this case.
+            entrypointRun.fail(
+                    new FailureModel(
+                            "Failed validating variables on start: " + exn.getMessage(),
+                            LHConstants.VAR_MUTATION_ERROR),
+                    now);
+            return;
+        }
+
+        for (ThreadVarDefModel threadVarDef : threadSpec.getVariableDefs()) {
+            VariableDefModel varDef = threadVarDef.getVarDef();
+            String varName = varDef.getName();
+            VariableValueModel val;
+
+            if (threadVarDef.getAccessLevel() == WfRunVariableAccessLevel.INHERITED_VAR) {
+                if (variables.containsKey(varName)) {
```

Comment:
This is confusing. This is the same as line 269.

Reply:
That's because this is handling inherited variables. For an inherited variable, the client who starts the `ThreadRun` shouldn't pass in a value for the inherited variable. That's because we use the shared/inherited one from the parent.

The reason why there is a `continue;` statement on line 266 is because we don't want to save a `Variable` to the rocksdb store. I will just add the validation in the `ThreadSpecModel#validateStartVariables()`